### PR TITLE
Issue #10086:Fix default AtclauseOrder and preserve order for documen…

### PR DESCRIPTION
**Pull request by Hariom-kr**
#10086 
- Created @PreserveOrder annotation to mark properties that should not be auto-sorted
- Applied @PreserveOrder to AtclauseOrderCheck.tagOrder field
- Updated XdocsPagesTest 
- Updated Javadoc

### CODE REVIEW

**Comment by Hariom-kr** [comment]
Github, generate website

**Comment by romani** [comment]
Xdoc files should be changed, to reflect actual order of tags. Generated website on this  PR should have new order of tags

**Comment by Hariom-kr** [comment]
Hi @romani , mvn clean verify,  is regenerating the meta XML file (`src/main/resources/.../AtclauseOrderCheck.xml`) back to the old alphabetical order. Should I update the meta generator?

**Comment by romani** [comment]
Yes , whole reason of issue is to make html page to show actual value.
We need a bit more changes, we are close to final fix, but need few more steps

**Comment by Hariom-kr** [comment]
Okay thanks, I am pushing the changes.

**Comment by romani** [comment]
GitHub, generate website

**Comment by Hariom-kr** [comment]
Github, generate website

**Comment by Hariom-kr** [comment]
Hi @romani , my local build passed with mvn clean verify, but CI is failing can u guide me on this?

**Comment by romani** [comment]
Please [read](https://github.com/checkstyle/checkstyle/blob/master/.github/CONTRIBUTING.md#getting-started) and watch videos at [Starting_Development](https://checkstyle.org/beginning_development.html#Starting_Development). Last video is about how to make CI green.
Please make CI green.

Checker, needs patches to be applied : https://github.com/checkstyle/checkstyle/actions/runs/18525806318/job/52796243069?pr=17918#step:7:11

fix: https://app.circleci.com/pipelines/github/checkstyle/checkstyle/37225/workflows/613e76ce-482e-4848-af67-958d5666480b/jobs/1101627/artifacts

**Comment by Hariom-kr** [comment]
Thanks for clarification , I understood How to fix this

File: config/import-control-test.xml [hunk 1]
```diff
@@ -43,6 +43,7 @@
   </subpackage>
 
   <subpackage name="internal">
+    <allow class="com.puppycrawl.tools.checkstyle.internal.annotation.PreserveOrder"/>
```
Comment (romani) [inline config/import-control-test.xml]: is this required? 
looks like we allow impot from same package

File: config/import-control.xml [hunk 2]
```diff
@@ -219,6 +221,11 @@
     </subpackage>
   </subpackage>
 
+  <subpackage name="internal">
+    <subpackage name="annotation">
+      <allow pkg="java.lang.annotation"/>
```
Comment (romani) [inline config/import-control.xml]: "java.lang.*" supposed to be completely allowed, as it is jdk classes and always available.
please recheck if this is allow actually required, if yes, lest allow all sub packages of java.lang.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AtclauseOrderCheck.java [hunk 3]
```diff
@@ -58,6 +59,7 @@ public class AtclauseOrderCheck extends AbstractJavadocCheck {
     /**
      * Default order of atclauses.
      */
+    @PreserveOrder
     private static final String[] DEFAULT_ORDER = {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AtclauseOrderCheck.java:62]: this might not be required as you put it on `private List<String> tagOrder`

**Comment by Hariom-kr** [comment]
Thank you for the review! Apologies, I'm still learning.. Fixing it now.

File: config/import-control.xml (hunk 2 above)
Comment (Hariom-kr) [reply config/import-control.xml]: PreserveOrder.java shows "Disallowed import" errors for java.lang.annotation classes , so allowed all subpackages of java.lang.

**Comment by Hariom-kr** [comment]
List of missing files in pitest profiles: src/main/java/com/puppycrawl/tools/checkstyle/internal/annotation/PreserveOrder.java this error is thrown on semaphore where to add preserveorder class?

**Comment by Hariom-kr** [comment]
@romani CI is happy now.

**Comment by Hariom-kr** [comment]
Github, generate website

**Comment by romani** [comment]
please respond to all my comments as "done" or any other disagreement or reason why not done.
Pleaase always do this, it helps me a lot.

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java [hunk 4]
```diff
@@ -1015,22 +1016,34 @@ private static String removeSquareBrackets(String value) {
...
+            // Check if field has @PreserveOrder annotation
+            final boolean preserveOrder = field != null
+                && field.isAnnotationPresent(PreserveOrder.class);
             try (Stream<?> valuesStream = getValuesStream(value)) {
-                result = valuesStream
+                final List<String> stringList = valuesStream
                     .map(String.class::cast)
+                    .collect(Collectors.toCollection(ArrayList<String>::new));
+
+                // Preserve original order or sort alphabetically
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: remove comments in this method that you added. Clean code does not require comments.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 5]
```diff
@@ -1513,10 +1515,19 @@ private static String getStringArrayPropertyValue(String propertyName, Object va
                 final Object[] array = (Object[]) value;
                 valuesStream = Arrays.stream(array);
             }
-            result = valuesStream
-                .map(String.class::cast)
-                .sorted()
-                .collect(Collectors.joining(", "));
+            // Check if field has @PreserveOrder annotation
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: remove comments that you added

File: pom.xml [hunk 6]
```diff
@@ -3376,6 +3376,28 @@
...
+          <plugin>
+            <groupId>org.pitest</groupId>
+            <artifactId>pitest-maven</artifactId>
+            <version>${pitest.plugin.version}</version>
+            <configuration>
+              <targetClasses>
+                <param>com.puppycrawl.tools.checkstyle.internal.annotation.*</param>
+              </targetClasses>
+              <coverageThreshold>0</coverageThreshold>
+              <mutationThreshold>0</mutationThreshold>
```
Comment (romani) [inline pom.xml]: please share details what forced  you todo this ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AtclauseOrderCheck.java (hunk 3 above)
Comment (Hariom-kr) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AtclauseOrderCheck.java:62]: Done

**Comment by Hariom-kr** [comment]
> please respond to all my comments as "done" or any other disagreement or reason why not done. Pleaase always do this, it helps me a lot.

Okay, Sure I will respond

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java [hunk 7]
```diff
@@ -947,7 +948,7 @@ else if (fieldClass == double[].class) {
             result = removeSquareBrackets(Arrays.toString((double[]) value).replace(".0", ""));
         }
         else if (fieldClass == String[].class) {
-            result = getStringArrayPropertyValue(value);
+            result = getStringArrayPropertyValue(value, field);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: please refactor this code to looks like
```
preserveOrder = hasPreserverOrderAnnoation(field);
getStringArrayPropertyValue(value, preserveOrder);
```

same for in other class.

File: config/import-control-test.xml (hunk 1 above)
Comment (Hariom-kr) [reply config/import-control-test.xml]: com.puppycrawl.tools.checkstyle.internal.* can access each other without explicit permission. Removed it.

File: pom.xml (hunk 6 above)
Comment (Hariom-kr) [reply pom.xml]: The CI failed with "List of missing files in pitest profiles" error for the new PreserveOrder.java file in internal.annotation package. I saw other packages like pitest-annotation and pitest-javadoc have their own pitest profiles in pom.xml, so I added pitest internalannotation to match that pattern and fix the build.
Screenshot 2025-10-18 232336
Also there is no logic test this is just a simple annotation so threshold set to zero..

Comment (romani) [reply pom.xml]: Can we add such class to any existing profile?
It just needs to be referenced in pom, nothing else.

Comment (romani) [reply pom.xml]: Please share link to CI failure, to let grap what job is doing this validation

Comment (Hariom-kr) [reply pom.xml]: Here is the Link: https://checkstyle.semaphoreci.com/workflows/60a02aaa-0fd7-4e1d-8606-1f392dc5e531?pipeline_id=8c930649-578b-4de3-8338-55d626ecd1aa

Comment (Hariom-kr) [reply pom.xml]: The internal.annotation package could be added to pitest-annotation (since it's an annotation) or if there's a general pitest-internal profile I missed

Comment (Hariom-kr) [reply pom.xml]: But PreserveOrder is just a marker annotation with no logic that needs testing.

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java (hunk 4 above)
Comment (Hariom-kr) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 5 above)
Comment (Hariom-kr) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java (hunk 7 above)
Comment (Hariom-kr) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: Done

File: pom.xml (hunk 6 above)
Comment (romani) [reply pom.xml]: `Validation (openjdk21, fast pool) - CMD=.ci/validation.sh check-missing-pitests logs`

you can run on local `.ci/validation.sh check-missing-pitests logs`
https://github.com/checkstyle/checkstyle/blob/a72e3f8fce30073691a42393f4f4bc2650a22cc5/.ci/validation.sh#L45-L51

just add this class here
https://github.com/checkstyle/checkstyle/blob/a72e3f8fce30073691a42393f4f4bc2650a22cc5/pom.xml#L3264-L3265

**Comment by Hariom-kr** [comment]
Some Internal errors occured, now its fine.

File: pom.xml (hunk 6 above)
Comment (Hariom-kr) [reply pom.xml]: okay, updating

Comment (Hariom-kr) [reply pom.xml]: Done

**Comment by Hariom-kr** [comment]
Github, generate website

**Comment by Hariom-kr** [comment]
@romani plz assign me a new issue to work on, i have finished all good xxxx issues..
